src/tds/convert.c

```diff
@@ -78,26 +78,37 @@ static const char *parse_numeric(const char *buf, const char *pend,
 
 #define test_alloc(x) {if ((x)==NULL) return TDS_CONVERT_NOMEM;}
 
-#define IS_TINYINT(x) ( 0 <= (x) && (x) <= 0xff )
-#define IS_SMALLINT(x) ( -32768 <= (x) && (x) <= 32767 )
-#define IS_USMALLINT(x) ( 0 <= (x) && (x) <= 65535 )
+/*
+ * Confirm convertability to a signed type with the given bounds,
+ * bearing in mind that promotion rules can complicate comparison
+ * between negative and unsigned integers, that conversion from
+ * floating-point values to integers starts by truncating fractional
+ * portions, and that lb - 1 might have the same FP representation as
+ * lb in x's type (as with float -> int and anything -> TDS_INT8).
+ */
+#define IS_INTn(x, lb, ub) ((x) < 0 ? lb - (x) < 1 : (x) < ub + 1)
+
+#define IS_TINYINT(x) ( 0 <= (x) && (x) < 0x100 )
+#define IS_SMALLINT(x) IS_INTn(x, -32768, 32767)
+#define IS_USMALLINT(x) ( 0 <= (x) && (x) < 65536 )
 /*
  * f77: I don't write -2147483648, some compiler seem to have some problem 
  * with this constant although is a valid 32bit value
  */
-#define TDS_INT_MIN (-2147483647l-1l)
-#define TDS_INT_MAX 2147483647l
-#define IS_INT(x) (TDS_INT_MIN <= (x) && (x) <= TDS_INT_MAX)
+#define TDS_INT_MIN (-2147483647-1)
+#define TDS_INT_MAX 2147483647
+#define IS_INT(x) IS_INTn(x, TDS_INT_MIN, (TDS_INT8) TDS_INT_MAX)
 
-#define TDS_UINT_MAX 4294967295lu
-#define IS_UINT(x) (0 <= (x) && (x) <= TDS_UINT_MAX)
+#define TDS_UINT_MAX 4294967295u
+#define IS_UINT(x) (0 <= (x) && (x) < (TDS_INT8) TDS_UINT_MAX + 1)
 
-#define TDS_INT8_MAX ((((TDS_INT8) 0x7fffffffl) << 32) + (TDS_INT8) 0xfffffffflu)
+#define TDS_INT8_MAX ((((TDS_INT8) 0x7fffffff) << 32) + (TDS_INT8) 0xffffffffu)
 #define TDS_INT8_MIN (-(((TDS_INT8)1)<<62) -(((TDS_INT8)1)<<62))
-#define IS_INT8(x) (TDS_INT8_MIN <= (x) && (x) <= TDS_INT8_MAX)
+#define IS_INT8(x) IS_INTn(x, TDS_INT8_MIN, (TDS_UINT8) TDS_INT8_MAX)
 
-#define TDS_UINT8_MAX ((((TDS_UINT8) 0xfffffffflu) << 32) + 0xfffffffflu)
+#define TDS_UINT8_MAX ((((TDS_UINT8) 0xffffffffu) << 32) + 0xffffffffu)
 #define IS_UINT8(x) (0 <= (x) && (x) <= TDS_UINT8_MAX)
+#define FLOAT_IS_UINT8(x) ((x) < -2.0f * TDS_INT8_MIN)
```

Comment:
I think you forgot to check for negativity.

Reply:
Oops, so I did.